pkg/osvscanner/vulnerability_result.go

```diff
@@ -11,17 +11,26 @@ import (
 	"github.com/google/osv-scanner/pkg/osv"
 )
 
+// Function used to filter vulnerabilities and groups.
+type vulnFilterFn func(sourcePath string, vulns []models.Vulnerability, groups []models.GroupInfo) ([]models.Vulnerability, []models.GroupInfo)
+
 // groupResponseBySource converts raw OSV API response into structured vulnerability information
 // grouped by source location.
-func groupResponseBySource(r *output.Reporter, query osv.BatchedQuery, resp *osv.HydratedBatchedResponse, callAnalysis bool) models.VulnerabilityResults {
+func groupResponseBySource(r *output.Reporter, query osv.BatchedQuery, resp *osv.HydratedBatchedResponse, callAnalysis bool, filter vulnFilterFn) models.VulnerabilityResults {
 	output := models.VulnerabilityResults{
 		Results: []models.PackageSource{},
 	}
 	groupedBySource := map[models.SourceInfo][]models.PackageVulns{}
 
+	totalFiltered := 0
+
 	for i, query := range query.Queries {
-		response := resp.Results[i]
-		if len(response.Vulns) == 0 {
+		vulns := resp.Results[i].Vulns
+		vulnCount := len(vulns)
+		groups := grouper.Group(grouper.ConvertVulnerabilityToIDAliases(vulns))
+		vulns, groups = filter(query.Source.Path, vulns, groups)
```

Comment:
Instead of passing a filter here, which can make the logic a bit hard to follow, can we just call a separate `filterIgnores` that takes in the entire `models.VulnerabilityResults` and modifies it as required?

Reply:
I've done a first pass at filtering the entire `models.VulnerabilityResults` - it's pretty deeply nested so I'll probably want to extract a few functions from it.

Reply:
Inner function has been extracted.